drivers/wifi/nrf700x/zephyr/src/zephyr_net_if.c

```diff
@@ -191,11 +191,9 @@ static void ip_maddr_event_handler(struct net_mgmt_event_callback *cb,
 		return;
```

Comment:
Do I understand correctly that this function is meant to only handle IP_MADDR events?

That is, the only events you are interested in are:

1. NET_EVENT_IPV4_MADDR_ADD
2. NET_EVENT_IPV4_MADDR_DEL
3. NET_EVENT_IPV6_MADDR_ADD
4. NET_EVENT_IPV6_MADDR_DEL

If so, then I would recommend adding a check to confirm that it really is one of these four events being raised. Currently, this event callback is being fired for all events in the `NET_EVENT_IPV4` and `NET_EVENT_IPV6` event groups.

This is because `net_mgmt` does not prevent event false positives, but only false negatives. So for instance:


```
net_mgmt_init_event_callback(&ip_maddr4_cb,
		     ip_maddr_event_handler,
		     NET_EVENT_IPV4_MADDR_ADD | NET_EVENT_IPV4_MADDR_DEL);
```

is guaranteed to fire for `NET_EVENT_IPV4_MADDR_ADD` and `NET_EVENT_IPV4_MADDR_DEL`, but it may also fire for any other `NET_EVENT_IPV4` event whose event code happens to fall within the bitmask

I think the full list of events that particular bitmask matches is

1. NET_EVENT_IPV4_CMD_ADDR_ADD
2. NET_EVENT_IPV4_CMD_ADDR_DEL
3. NET_EVENT_IPV4_CMD_MADDR_ADD
4. NET_EVENT_IPV4_CMD_MADDR_DEL
5. NET_EVENT_IPV4_CMD_ROUTER_ADD
6. NET_EVENT_IPV4_CMD_ROUTER_DEL
7. NET_EVENT_IPV4_CMD_DHCP_START


See the network management docs for more details:
https://docs.zephyrproject.org/latest/connectivity/networking/api/net_mgmt.html

Namely:

> Event set filtering allows false positives for events that have the same layer and layer code. A callback handler function must check the event code (passed as an argument) against the specific network events it will handle, regardless of how many events were in the set passed to [net_mgmt_init_event_callback()](https://docs.zephyrproject.org/latest/connectivity/networking/api/net_mgmt.html#c.net_mgmt_init_event_callback).

Reply:
I did see those two if statements, but consider for example if `NET_EVENT_IPV4_CMD_DHCP_START` is fired

`wifi_nrf_fmac_set_mcast_addr` will still fire even though `NET_EVENT_IPV4_CMD_DHCP_START` is not one of the four events you are interested in, because those if statements serve only to configure the arguments that will be passed to it.

You need, for instance, an additional else statement which returns out of the event handler if it is none of the four target events

Reply:
You're right, return early makes way more sense -- either way, you need something which actually cancels handling the event if it's not one of the four target events

Reply:
Also, there is a memory leak, so, fixing that as well :)